drivers/soundwire/intel_init.c

```diff
@@ -301,17 +379,35 @@ static struct sdw_intel_ctx
 
 	return ctx;
 
-err:
-	ctx->count = i;
-	sdw_intel_cleanup(ctx);
+	/* the loop counters are different for the init and add device loops,
+	 * which makes it easier to deal with the error unwinding paths.
+	 */
+del:
+	while (j--) {
+		if (!(link_mask & BIT(j)))
+			continue;
+		ldev = ctx->ldev[j];
+		auxiliary_device_delete(&ldev->auxdev);
+	}
+
+uninit:
+	while (i--) {
+		if (!(link_mask & BIT(i)))
+			continue;
+		ldev = ctx->ldev[i];
+		auxiliary_device_uninit(&ldev->auxdev);
```

Comment:
is a kfree(ctx->ldev[i]) missing here?

Reply:
uh no. this is good.

Reply:
@bardliao after the init stage, the memory is freed by the put_device that will result in the .release function being called.
So the memory is freed by the device core, we don't need to add a manual kfree.

Reply:
@bardliao yeah this is not easy to see but the explanation is also there on line 314 above.

Reply:
Thanks @plbossart and @ranj063 I got it now.